prow/cmd/artifact-uploader/main.go

```diff
@@ -0,0 +1,166 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/clientcmd"
+	"k8s.io/test-infra/prow/kube"
+
+	"k8s.io/test-infra/prow/artifact-uploader"
+	"k8s.io/test-infra/prow/gcsupload"
+	"k8s.io/test-infra/prow/logrusutil"
+	"k8s.io/test-infra/prow/pod-utils/options"
+)
+
+// NewOptions returns an empty Options with no nil fields
+func NewOptions() *Options {
+	return &Options{
+		Options: gcsupload.NewOptions(),
+	}
+}
+
+type Options struct {
+	// NumWorkers determines the number of workers that consume
+	// the controller's work queue
+	NumWorkers int `json:"num_workers"`
+	// ProwJobNamespace is the namespace in the cluster that holds
+	// ProwJob objects
+	ProwJobNamespace string `json:"'prow_job_namespace'"`
+
+	*gcsupload.Options
+
+	clusterConfig *rest.Config
+}
+
+// Validate ensures that the set of options are
+// self-consistent and valid
+func (o *Options) Validate() error {
+	if o.NumWorkers == 0 {
+		return errors.New("number of workers cannot be zero")
+	}
+
+	if o.ProwJobNamespace == "" {
+		return errors.New("namespace containing ProwJobs not configured")
+	}
+
+	return o.Options.Validate()
+}
+
+const (
+	// JSONConfigEnvVar is the environment variable that
+	// utilities expect to find a full JSON configuration
+	// in when run.
+	JSONConfigEnvVar = "ARTIFACTUPLOAD_OPTIONS"
+)
+
+// ConfigVar exposes the environment variable used
+// to store serialized configuration
+func (o *Options) ConfigVar() string {
+	return JSONConfigEnvVar
+}
+
+// LoadConfig loads options from serialized config
+func (o *Options) LoadConfig(config string) error {
+	return json.Unmarshal([]byte(config), o)
+}
+
+// BindOptions binds flags to options
+func (o *Options) BindOptions(flags *flag.FlagSet) {
+	flags.IntVar(&o.NumWorkers, "num-workers", 25, "Number of threads to use for processing updates.")
+	flags.StringVar(&o.ProwJobNamespace, "prow-job-ns", "", "Namespace containing ProwJobs.")
+	gcsupload.BindOptions(o.Options, flags)
+}
+
+// Complete internalizes command line arguments
+func (o *Options) Complete(args []string) {
+	o.Options.Complete(args)
+}
+
+// loadClusterConfig loads connection configuration
+// for the cluster we're deploying to. We prefer to
+// use in-cluster configuration if possible, but will
+// fall back to using default rules otherwise.
+func loadClusterConfig() (*rest.Config, error) {
```

Comment:
This tries in-cluster first, if that does not work it falls back to other methods (`$KUBECONFIG`, `~/.kube/config`, etc). We can make it explicit later, like you say, if we don't want the controller deployed on the cluster with the ProwJobs or test Pods.

Reply:
Ok, so this won't work with separate build clusters as is because the in-cluster config for the service cluster will be found?
When would we fail to find in-cluster config?